### PR TITLE
[Map] Rework interaction (entries, keys, values, extract) of NULL MAPs

### LINKED ISSUE #11116: `MAP_KEYS()`, `MAP_VALUES()`, `MAP_ENTRIES()` do not propagate dtypes with NULLs
### What happens?

```sql
SELECT TYPEOF(MAP_KEYS(NULL::MAP(TEXT, BIGINT)));
--- gives NULL, should be TEXT[]
SELECT TYPEOF(MAP_VALUES(NULL::MAP(TEXT, BIGINT)));
--- gives NULL, should be BIGINT[]
SELECT TYPEOF(MAP_ENTRIES(NULL::MAP(TEXT, BIGINT)));
--- gives NULL, should be list<struct<key: TEXT, value: BIGINT>>
SELECT TYPEOF(MAP_EXTRACT(NULL::MAP(TEXT, BIGINT), 'a'));
--- gives BIGINT[], as expected
```

### To Reproduce

see above

### OS:

shell.duckdb.org

### DuckDB Version:

0.10.0

### DuckDB Client:

@duckdb/duckdb-wasm@1.28.1-dev166.0

### Full Name:

Nick Crews

### Affiliation:

Ship Creek Group

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #11116 

### OId behavior:
`TYPEOF(MAP_KEYS(NULL::MAP(TEXT, BIGINT)))` -> `NULL`
`TYPEOF(MAP_VALUES(NULL::MAP(TEXT, BIGINT)))` -> `NULL`
`TYPEOF(MAP_ENTRIES(NULL::MAP(TEXT, BIGINT)))` -> `NULL`
`TYPEOF(MAP_EXTRACT(NULL::MAP(TEXT, BIGINT), 'a'))` -> `BIGINT[]`

`TYPEOF(MAP_KEYS(NULL))` -> `NULL`
`TYPEOF(MAP_VALUES(NULL))` -> `NULL`
`TYPEOF(MAP_ENTRIES(NULL))` -> `NULL`
`TYPEOF(MAP_EXTRACT(NULL, 'a'))` -> `Binder Error: MAP_EXTRACT can only operate on MAPs`

### New behavior:
`TYPEOF(MAP_KEYS(NULL::MAP(TEXT, BIGINT)))` -> `VARCHAR[] `
`TYPEOF(MAP_VALUES(NULL::MAP(TEXT, BIGINT)))` -> `BIGINT[]`
`TYPEOF(MAP_ENTRIES(NULL::MAP(TEXT, BIGINT)))` -> `STRUCT("key" VARCHAR, "value" BIGINT)[]`
`TYPEOF(MAP_EXTRACT(NULL::MAP(TEXT, BIGINT), 'a'))` -> `BIGINT[]`

`TYPEOF(MAP_KEYS(NULL))` -> `NULL[]`
`TYPEOF(MAP_VALUES(NULL))` -> `NULL[]`
`TYPEOF(MAP_ENTRIES(NULL))` -> `STRUCT("key" NULL, "value" NULL)[]`
`TYPEOF(MAP_EXTRACT(NULL, 'a'))` -> `NULL[]`

Note:
`MAP_EXTRACT` returns a `<type>[]` and will never return NULL.
If the entry is not present the result is `[]` (empty list)
This behavior originated from Presto behavior when the method was first implemented.

Ideally we change this to `<type>`, returning NULL if the entry is not present, as this matches the behavior of other systems.

This is quite a big change in behavior that will affect every use of `map_extract` and also the shorthand `map_column[...]` so that will be a separate change if we even want to make this breaking change.

### CODE REVIEW

**Comment by Tishj** [comment]
@Mytherin can you rerun <https://github.com/duckdb/duckdb/actions/runs/8764095759/job/24053721718?pr=11745>?
It's a Python regression test failure, seems unrelated

**Comment by Tishj** [comment]
@nickcrews

> What should `TYPEOF(MAP_KEYS(MAP(NULL, [1,2,3])))` be?

That would be `NULL[]`
`MAP(NULL, [1,2,3])` produces a NULL value

`MAP_KEYS(NULL)` -> `NULL[]`

**Comment by NickCrews** [comment]
Thank @Tishj , I think that makes sense. I was mostly asking rhetorically, I wasn't sure if we should even allow someone to construct a MAP with a plain uncasted NULL. But if there is precedence throughout the rest of duckdb to have NULLTYPE then that sounds good. Thank you!

File: test/sql/types/nested/map/test_map_entries.test [hunk 1]
```diff
@@ -82,3 +82,30 @@ SELECT map_entries(map_from_entries(
...
+select MAP_ENTRIES(MAP(NULL, NULL))
+----
+NULL
+
+query I
+select MAP_ENTRIES(NULL)
+----
+NULL
+
+# FIXME: depends on #11625 for "NULL" ToString support
```
Comment (Mytherin) [inline test/sql/types/nested/map/test_map_entries.test:101]: Could you address these fixmes? #11625 has been merged now

**Review by Mytherin — COMMENTED** [review]
Thanks! LGTM - one minor comment:

File: src/core_functions/scalar/map/map_entries.cpp [hunk 2]
```diff
@@ -36,25 +55,24 @@ static unique_ptr<FunctionData> MapEntriesBind(ClientContext &context, ScalarFun
...
-	bound_function.return_type = LogicalType::LIST(row_type);
+	bound_function.return_type = CreateReturnType(map);
 	return make_uniq<VariableReturnBindData>(bound_function.return_type);
 }
 
 ScalarFunction MapEntriesFun::GetFunction() {
 	//! the arguments and return types are actually set in the binder function
 	ScalarFunction fun({}, LogicalTypeId::LIST, MapEntriesFunction, MapEntriesBind);
-	fun.null_handling = FunctionNullHandling::DEFAULT_NULL_HANDLING;
+	fun.null_handling = FunctionNullHandling::SPECIAL_HANDLING;
```
Comment (Mytherin) [inline src/core_functions/scalar/map/map_entries.cpp:75]: These functions should not use special handling - this seems like a work-around for another work-around in `src/function/function_binder.cpp` where we don't return the output type but convert it to `SQLNULL` as a fallback.

Can we apply this diff to fix that, and turn this back into `DEFAULT_NULL_HANDLING`

```diff
diff --git a/src/function/function_binder.cpp b/src/function/function_binder.cpp
index 267f589d8f..87eacee685 100644
--- a/src/function/function_binder.cpp
+++ b/src/function/function_binder.cpp
@@ -306,13 +306,20 @@ unique_ptr<Expression> FunctionBinder::BindScalarFunction(ScalarFunctionCatalogE
        LogicalType return_type_if_null;
        switch (bound_function.return_type.id()) {
        case LogicalTypeId::ANY:
+               return_type_if_null = LogicalType::SQLNULL;
+               break;
        case LogicalTypeId::DECIMAL:
        case LogicalTypeId::STRUCT:
        case LogicalTypeId::LIST:
        case LogicalTypeId::MAP:
        case LogicalTypeId::UNION:
        case LogicalTypeId::ARRAY:
-               return_type_if_null = LogicalType::SQLNULL;
+               if (bound_function.return_type.AuxInfo()) {
+                       // this is a complete and valid type - return it
+                       return_type_if_null = bound_function.return_type;
+               } else {
+                       return_type_if_null = LogicalType::SQLNULL;
+               }
                break;
        default:
                return_type_if_null = bound_function.return_type;

```

File: src/core_functions/scalar/map/map_entries.cpp [hunk 3]
```diff
@@ -12,18 +12,37 @@ namespace duckdb {
 static void MapEntriesFunction(DataChunk &args, ExpressionState &state, Vector &result) {
 	auto count = args.size();
 
-	MapUtil::ReinterpretMap(result, args.data[0], count);
+	auto &map = args.data[0];
+	if (map.GetType().id() == LogicalTypeId::SQLNULL) {
+		// Input is a constant NULL
+		auto &validity = FlatVector::Validity(result);
+		validity.SetInvalid(0);
+		result.SetVectorType(VectorType::CONSTANT_VECTOR);
```
Comment (Mytherin) [inline src/core_functions/scalar/map/map_entries.cpp:20]: Probably we won't need this anymore either after we have `DEFAULT_NULL_HANDLING`

**Review by Mytherin — COMMENTED** [review]
One more comment, I think this can be done a bit more cleanly:

File: src/core_functions/scalar/map/map_entries.cpp (hunk 2 above)
Comment (Tishj) [reply src/core_functions/scalar/map/map_entries.cpp:75]:
```
Wrong result in query! (test/sql/types/nested/map/test_null_map_interaction.test:27)!
================================================================================
SELECT TYPEOF(MAP_ENTRIES(NULL::MAP(TEXT, BIGINT)));
================================================================================
Mismatch on row 1, column 1
"NULL" <> STRUCT("key" VARCHAR, "value" BIGINT)[]
================================================================================
Expected result:
================================================================================
STRUCT("key" VARCHAR, "value" BIGINT)[]
================================================================================
Actual result:
================================================================================
"NULL"
```

Looks like that's not enough, I'll look into it

Comment (Tishj) [reply src/core_functions/scalar/map/map_entries.cpp:75]: This kicks in before we get to the `bind` function of the ScalarFunction
It happens after figuring out which overload to use, but before we run Bind

Comment (Mytherin) [reply src/core_functions/scalar/map/map_entries.cpp:75]: I see - I suppose the nice solution here would be to wait for @Maxxen's rework on how type binding is handled as that would perhaps solve this, but we can merge this before that lands

Comment (Tishj) [reply src/core_functions/scalar/map/map_entries.cpp:75]: @Mytherin so should we merge this as-is then?
I think all the comments were related to DEFAULT_NULL_HANDLING

Comment (Mytherin) [reply src/core_functions/scalar/map/map_entries.cpp:75]: Yes that's fine

**Comment by Mytherin** [comment]
Could you have a look at the failing CI? Otherwise LGTM

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**